### PR TITLE
Fix for #8933 - Copy and paste file/folder in the same place will duplicate the file/folder

### LINKED ISSUE #8933: Copy & Paste issues
### Bug Description:

Issue 1
When user select file or folder and press 'Ctrl+C' and 'Ctrl+V' he error.
At vscode a new file created that end with copy suffix.
This functionality is exist when user select duplicate from the menu, can it be used in the same way here?

Issue 2
Select paste at chrome browser from the menu is not supported and return this error https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/common-frontend-contribution.ts#L573
Can it be supported? 
If it can't be supported, can the command not be registered (vscode doesn't registered unsupported commands).

### Additional Information

- Operating System: Gitpod
- Theia Version: last version

### PR DESCRIPTION
**Pull request by ronabet**
Signed-off-by: Ron Nabet <ron.nabet@sap.com>

#### What it does

Fixes #8933
Select file/folder and press 'Ctrl+C' and 'Ctrl+V' and the file/folder will be duplicated instead of an error.

#### How to test

Mark file/folder and press 'Ctrl+C' and 'Ctrl+V'.
Copy a file/folder and paste under the same folder, the file will be created with "_copy" at the end of the filename.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by kittaakos** [comment]
Please do not change `FileService`, it was [copied from VS Code](https://github.com/microsoft/vscode/blob/master/src/vs/platform/files/common/fileService.ts#L653-L709). In fact, I'd move out the Theia-specific code from the `FileService` when generating a non-conflicting basename. Please change.


Here is a working proposal, it's up to you how you implement it, but we won't hack the `FileService`. Thank you for your understanding.
```diff
diff --git a/packages/filesystem/src/browser/file-service.ts b/packages/filesystem/src/browser/file-service.ts
index 89493b5e5..2f3b3918b 100644
--- a/packages/filesystem/src/browser/file-service.ts
+++ b/packages/filesystem/src/browser/file-service.ts
@@ -1117,13 +1117,6 @@ export class FileService {
         // validation
         const { exists, isSameResourceWithDifferentPathCase } = await this.doValidateMoveCopy(sourceProvider, source, targetProvider, target, mode, overwrite);
 
-        // if target exists get valid target
-        if (exists && !overwrite) {
-            const parent = await this.resolve(target.parent);
-            const name = target.path.name + '_copy';
-            target = FileSystemUtils.generateUniqueResourceURI(target.parent, parent, name, target.path.ext);
-        }
-
         // delete as needed (unless target is same resource with different path case)
         if (exists && !isSameResourceWithDifferentPathCase && overwrite) {
             await this.delete(target, { recursive: true });
diff --git a/packages/filesystem/src/browser/file-tree/file-tree-model.ts b/packages/filesystem/src/browser/file-tree/file-tree-model.ts
index 1f8440e41..38a128c25 100644
--- a/packages/filesystem/src/browser/file-tree/file-tree-model.ts
+++ b/packages/filesystem/src/browser/file-tree/file-tree-model.ts
@@ -24,6 +24,7 @@ import { FileService } from '../file-service';
 import { FileOperationError, FileOperationResult, FileChangesEvent, FileChangeType, FileChange, FileOperation, FileOperationEvent } from '../../common/files';
 import { MessageService } from '@theia/core/lib/common/message-service';
 import { EnvVariablesServer } from '@theia/core/lib/common/env-variables';
+import { FileSystemUtils } from '../../common';
 
 @injectable()
 export class FileTreeModel extends TreeModelImpl implements LocationService {
@@ -148,8 +149,14 @@ export class FileTreeModel extends TreeModelImpl implements LocationService {
     }
 
     async copy(source: URI, target: Readonly<FileStatNode>): Promise<URI> {
-        const targetUri = target.uri.resolve(source.path.base);
+        const baseUri = (DirNode.is(target) ? target : ((target.parent || this.root!) as Readonly<FileStatNode>)).fileStat.resource;
+        let targetUri = baseUri.resolve(source.path.base);
         try {
+            if (source.toString() === targetUri.toString()) {
+                const parent = await this.fileService.resolve(targetUri.parent);
+                const name = targetUri.path.name + '_copy';
+                targetUri = FileSystemUtils.generateUniqueResourceURI(targetUri.parent, parent, name, targetUri.path.ext);
+            }
             await this.fileService.copy(source, targetUri);
         } catch (e) {
             this.messageService.error(e.message);
```

**Comment by ronabet** [comment]
> Please do not change `FileService`, it was [copied from VS Code](https://github.com/microsoft/vscode/blob/master/src/vs/platform/files/common/fileService.ts#L653-L709). In fact, I'd move out the Theia-specific code from the `FileService` when generating a non-conflicting basename. Please change.
> 
> Here is a working proposal, it's up to you how you implement it, but we won't hack the `FileService`. Thank you for your understanding.
> 
> ```diff
> diff --git a/packages/filesystem/src/browser/file-service.ts b/packages/filesystem/src/browser/file-service.ts
> index 89493b5e5..2f3b3918b 100644
> --- a/packages/filesystem/src/browser/file-service.ts
> +++ b/packages/filesystem/src/browser/file-service.ts
> @@ -1117,13 +1117,6 @@ export class FileService {
>          // validation
>          const { exists, isSameResourceWithDifferentPathCase } = await this.doValidateMoveCopy(sourceProvider, source, targetProvider, target, mode, overwrite);
>  
> -        // if target exists get valid target
> -        if (exists && !overwrite) {
> -            const parent = await this.resolve(target.parent);
> -            const name = target.path.name + '_copy';
> -            target = FileSystemUtils.generateUniqueResourceURI(target.parent, parent, name, target.path.ext);
> -        }
> -
>          // delete as needed (unless target is same resource with different path case)
>          if (exists && !isSameResourceWithDifferentPathCase && overwrite) {
>              await this.delete(target, { recursive: true });
> diff --git a/packages/filesystem/src/browser/file-tree/file-tree-model.ts b/packages/filesystem/src/browser/file-tree/file-tree-model.ts
> index 1f8440e41..38a128c25 100644
> --- a/packages/filesystem/src/browser/file-tree/file-tree-model.ts
> +++ b/packages/filesystem/src/browser/file-tree/file-tree-model.ts
> @@ -24,6 +24,7 @@ import { FileService } from '../file-service';
>  import { FileOperationError, FileOperationResult, FileChangesEvent, FileChangeType, FileChange, FileOperation, FileOperationEvent } from '../../common/files';
>  import { MessageService } from '@theia/core/lib/common/message-service';
>  import { EnvVariablesServer } from '@theia/core/lib/common/env-variables';
> +import { FileSystemUtils } from '../../common';
>  
>  @injectable()
>  export class FileTreeModel extends TreeModelImpl implements LocationService {
> @@ -148,8 +149,14 @@ export class FileTreeModel extends TreeModelImpl implements LocationService {
>      }
>  
>      async copy(source: URI, target: Readonly<FileStatNode>): Promise<URI> {
> -        const targetUri = target.uri.resolve(source.path.base);
> +        const baseUri = (DirNode.is(target) ? target : ((target.parent || this.root!) as Readonly<FileStatNode>)).fileStat.resource;
> +        let targetUri = baseUri.resolve(source.path.base);
>          try {
> +            if (source.toString() === targetUri.toString()) {
> +                const parent = await this.fileService.resolve(targetUri.parent);
> +                const name = targetUri.path.name + '_copy';
> +                targetUri = FileSystemUtils.generateUniqueResourceURI(targetUri.parent, parent, name, targetUri.path.ext);
> +            }
>              await this.fileService.copy(source, targetUri);
>          } catch (e) {
>              this.messageService.error(e.message);
> ```

@kittaakos Can you see again?

**Review by kittaakos — COMMENTED** [review]
First off, your PR is not the same as my proposal. Secondly, please squash your commits. Finally, I won't review and accept my code, it's against our review principles. Please find someone else for the review. Thank you!

**Comment by ronabet** [comment]
> First off, your PR is not the same as my proposal. Secondly, please squash your commits. Finally, I won't review and accept my code, it's against our review principles. Please find someone else for the review. Thank you!

@kittaakos 
I did exactly according to your proposal, can you direct me to specific lines that do not same?

**Comment by ronabet** [comment]
> First off, your PR is not the same as my proposal. Secondly, please squash your commits. Finally, I won't review and accept my code, it's against our review principles. Please find someone else for the review. Thank you!

Fixed

**Comment by amiramw** [comment]
@ronabet when I try to Ctrl+C and then Ctrl+V on a folder I get infinite nested libraries. In your description it says folder copy should be fixed as well in this PR.

**Comment by ronabet** [comment]
> @ronabet when I try to Ctrl+C and then Ctrl+V on a folder I get infinite nested libraries. In your description it says folder copy should be fixed as well in this PR.

Fixed

File: packages/filesystem/src/browser/file-tree/file-tree-model.ts [hunk 1]
```diff
@@ -148,8 +149,13 @@ export class FileTreeModel extends TreeModelImpl implements LocationService {
     }
 
     async copy(source: URI, target: Readonly<FileStatNode>): Promise<URI> {
-        const targetUri = target.uri.resolve(source.path.base);
+        let targetUri = target.uri.resolve(source.path.base);
         try {
+            if (source.path.toString() === target.uri.path.toString()) {
+            const parent = await this.fileService.resolve(source.parent);
```
Comment (amiramw) [inline packages/filesystem/src/browser/file-tree/file-tree-model.ts:155]: something is wrong with the formatting

Comment (ronabet) [reply packages/filesystem/src/browser/file-tree/file-tree-model.ts:155]: can you explain?

Comment (amiramw) [reply packages/filesystem/src/browser/file-tree/file-tree-model.ts:155]: there is no indentation for the if block. VS Code should do formatting on save.

Comment (ronabet) [reply packages/filesystem/src/browser/file-tree/file-tree-model.ts]: the `if` indentation fixed.
formatting happens in the same way in duplicate, see this: https://github.com/eclipse-theia/theia/blob/57b4b671f621dc1c5c6eec7f8c6ab1fe0585e4a5/packages/workspace/src/browser/workspace-duplicate-handler.ts#L67

**Comment by amiramw** [comment]
@ronabet @i501378 is it intentional that now https://github.com/eclipse-theia/theia/pull/8778 doesn't work? Instead of pasting the file with suffix I see now error notification "File at target already exists".

**Comment by ronabet** [comment]
> @ronabet @i501378 is it intentional that now #8778 doesn't work? Instead of pasting the file with suffix I see now error notification "File at target already exists".

I fixed it.

**Review by amiramw — APPROVED** [review]

**Comment by kittaakos** [comment]
FYI: you have merged a PR that was different than the expected one. The `FileService` contains the same logic you have added. 
https://github.com/eclipse-theia/theia/blob/b972e2b217da074d821d173a4f8517e171efbd45/packages/filesystem/src/browser/file-service.ts#L1121-L1125

Related: https://github.com/eclipse-theia/theia/pull/8940#discussion_r572939690

**Comment by amiramw** [comment]
Oh, sorry. I think when @ronabet removed the code from file-service then https://github.com/eclipse-theia/theia/pull/8778 flow got broken.
How do you suggest to proceed?

**Comment by kittaakos** [comment]
> How do you suggest to proceed?

OK, no worries. I open a follow-up and will take care of it.

**Pull request merged**